File: python/chatbot_demo.py
#!/usr/bin/env python3
"""
AI Chatbot Demo with Compliance Integration
Integrates with Brev LLM for intelligent responses and compliance checking
"""

import json
import subprocess
import sys
import time
from typing import Dict, List, Optional
import requests
from datetime import datetime
import os

class ChatbotComplianceDemo:
    def __init__(self):
        self.session_id = f"session_{int(time.time())}"
        self.user_id = "demo_user"
        self.conversation_history = []
        self.compliance_violations = []
        
        # Brev configuration
        self.brev_api_base = os.getenv("BREV_API_BASE", "https://api.brev.com/v1")
        self.brev_api_key = os.getenv("BREV_API_KEY", "")
        self.model_name = "llama-3.3-nemotron-super-49b-v1"
        
        # Compliance system path
        self.rust_binary = "../target/debug/pii-compliance-agent"
        
    def call_brev_llm(self, prompt: str, max_tokens: int = 150) -> str:
        """Call Brev LLM for intelligent responses"""
        if not self.brev_api_key:
            return self._get_mock_response(prompt)
            
        headers = {
            "Authorization": f"Bearer {self.brev_api_key}",
            "Content-Type": "application/json"
        }
        
        payload = {
            "model": self.model_name,
            "messages": [
                {
                    "role": "system",
                    "content": "You are a helpful customer service chatbot. Keep responses concise and professional."
                },
                {
                    "role": "user", 
                    "content": prompt
                }
            ],
            "max_tokens": max_tokens,
            "temperature": 0.7
        }
        
        try:
            response = requests.post(
                f"{self.brev_api_base}/chat/completions",
                headers=headers,
                json=payload,
                timeout=30
            )
            
            if response.status_code == 200:
                result = response.json()
                return result["choices"][0]["message"]["content"].strip()
            else:
                print(f"⚠️  Brev API error: {response.status_code}")
                return self._get_mock_response(prompt)
                
        except Exception as e:
            print(f"⚠️  Brev API call failed: {e}")
            return self._get_mock_response(prompt)
    
    def _get_mock_response(self, prompt: str) -> str:
        """Mock responses when Brev is not available"""
        prompt_lower = prompt.lower()
        
        if "email" in prompt_lower or "contact" in prompt_lower:
            return "I'd be happy to help! Could you please provide your email address so I can assist you better?"
        elif "account" in prompt_lower:
            return "I can help you with your account. What specific issue are you experiencing?"
        elif "payment" in prompt_lower or "card" in prompt_lower:
            return "I can help you update your payment information. What would you like to change?"
        elif "appointment" in prompt_lower or "schedule" in prompt_lower:
            return "I can help you schedule an appointment. What type of service do you need?"
        elif "help" in prompt_lower:
            return "I'm here to help! What can I assist you with today?"
        else:
            return "Thank you for your message. How can I help you today?"
    
    def check_compliance(self, message: str) -> Dict:
        """Check message compliance using Rust binary"""
        try:
            # Call the Rust compliance system
            result = subprocess.run(
                [self.rust_binary, "scan", "--text", message],
                capture_output=True,
                text=True,
                cwd=".."
            )
            
            if result.returncode == 0:
                # Parse the JSON output from Rust
                output_lines = result.stdout.strip().split('\n')
                json_start = output_lines.index("📋 Compliance Report:") + 1
                json_str = '\n'.join(output_lines[json_start:])
                
                try:
                    compliance_data = json.loads(json_str)
                    return compliance_data
                except json.JSONDecodeError:
                    return self._parse_compliance_output(result.stdout)
            else:
                print(f"⚠️  Compliance check failed: {result.stderr}")
                return self._get_mock_compliance_result(message)
                
        except Exception as e:
            print(f"⚠️  Compliance system error: {e}")
            return self._get_mock_compliance_result(message)
    
    def _parse_compliance_output(self, output: str) -> Dict:
        """Parse compliance output when JSON parsing fails"""
        lines = output.split('\n')
        compliance_data = {
            "original_text": "",
            "redacted_text": "",
            "detected_pii": [],
            "compliance_score": 1.0,
            "recommendations": []
        }
        
        for line in lines:
            if "Original:" in line:
                compliance_data["original_text"] = line.split("Original:", 1)[1].strip()
            elif "Redacted:" in line:
                compliance_data["redacted_text"] = line.split("Redacted:", 1)[1].strip()
            elif "Score:" in line:
                try:
                    score_str = line.split("Score:", 1)[1].strip().replace("%", "")
                    compliance_data["compliance_score"] = float(score_str) / 100.0
                except:
                    pass
            elif "PII Found:" in line:
                try:
                    pii_count = int(line.split("PII Found:", 1)[1].strip().split()[0])
                    if pii_count > 0:
                        compliance_data["compliance_score"] = max(0.1, compliance_data["compliance_score"] - 0.3)
                except:
                    pass
        
        return compliance_data
    
    def _get_mock_compliance_result(self, message: str) -> Dict:
        """Mock compliance result for testing"""
        return {
            "original_text": message,
            "redacted_text": message,
            "detected_pii": [],
            "compliance_score": 1.0,
            "recommendations": ["✅ Message appears compliant"]
        }
    
    def log_conversation(self, user_message: str, bot_response: str, compliance_result: Dict):
        """Log conversation with compliance data"""
        entry = {
            "timestamp": datetime.now().isoformat(),
            "user_message": user_message,
            "bot_response": bot_response,
            "compliance_score": compliance_result.get("compliance_score", 1.0),
            "pii_detected": len(compliance_result.get("detected_pii", [])),
            "redacted_text": compliance_result.get("redacted_text", user_message)
        }
        
        self.conversation_history.append(entry)
        
        # Check for compliance violations
        if compliance_result.get("compliance_score", 1.0) < 0.8:
            violation = {
                "timestamp": datetime.now().isoformat(),
                "severity": "HIGH" if compliance_result.get("compliance_score", 1.0) < 0.6 else "MEDIUM",
                "pii_count": len(compliance_result.get("detected_pii", [])),
                "original_message": user_message,
                "redacted_message": compliance_result.get("redacted_text", user_message)
            }
            self.compliance_violations.append(violation)
    
    def run_demo(self):
        """Run the interactive chatbot demo"""
        print("🤖 AI Chatbot Compliance Demo")
        print("=" * 50)
        print(f"Session ID: {self.session_id}")
        print(f"User ID: {self.user_id}")
        print("Type 'quit' to exit, 'help' for commands")
        print("=" * 50)
        
        # Demo scenarios
        demo_scenarios = [
            "Hi, I need help with my account",
            "My email is john.doe@company.com",
            "My phone number is 555-123-4567",
            "My SSN is 123-45-6789",
            "My credit card is 4111-1111-1111-1111",
            "What's the weather like today?"
        ]
        
        scenario_index = 0
        
        while True:
            try:
                if scenario_index < len(demo_scenarios):
                    # Auto-demo mode
                    user_input = demo_scenarios[scenario_index]
                    print(f"\n👤 User: {user_input}")
                    time.sleep(1)
                else:
                    # Interactive mode
                    user_input = input("\n👤 You: ").strip()
                
                if user_input.lower() in ['quit', 'exit', 'q']:
                    break
                elif user_input.lower() == 'help':
                    self._show_help()
                    continue
                elif user_input.lower() == 'status':
                    self._show_status()
                    continue
                elif user_input.lower() == 'violations':
                    self._show_violations()
                    continue
                
                # Check compliance BEFORE processing
                print("🔍 Checking compliance...")
                compliance_result = self.check_compliance(user_input)
                
                # Generate bot response using Brev LLM
                print("🤖 Generating response...")
                bot_response = self.call_brev_llm(user_input)
                
                # Log conversation
                self.log_conversation(user_input, bot_response, compliance_result)
                
                # Display results
                print(f"🤖 Bot: {bot_response}")
                print(f"📊 Compliance Score: {compliance_result.get('compliance_score', 1.0):.1%}")
                
                if compliance_result.get("detected_pii"):
                    print(f"⚠️  PII Detected: {len(compliance_result['detected_pii'])} items")
                    for pii in compliance_result["detected_pii"]:
                        print(f"   - {pii.get('pii_type', 'Unknown')}: {pii.get('value', 'N/A')}")
                
                if compliance_result.get("redacted_text") != user_input:
                    print(f"🔒 Redacted: {compliance_result['redacted_text']}")
                
                # Show recommendations
                if compliance_result.get("recommendations"):
                    print("💡 Recommendations:")
                    for rec in compliance_result["recommendations"][:3]:  # Show first 3
                        print(f"   - {rec}")
                
                scenario_index += 1
                
            except KeyboardInterrupt:
                print("\n\n👋 Demo ended by user")
                break
            except Exception as e:
                print(f"❌ Error: {e}")
                continue
        
        # Show final summary
        self._show_final_summary()
    
    def _show_help(self):
        """Show available commands"""
        print("\n📋 Available Commands:")
        print("  help        - Show this help")
        print("  status      - Show conversation status")
        print("  violations  - Show compliance violations")
        print("  quit        - Exit demo")
        print("\n💡 Demo will auto-run scenarios, then switch to interactive mode")
    
    def _show_status(self):
        """Show conversation status"""
        print(f"\n📊 Conversation Status:")
        print(f"  Messages: {len(self.conversation_history)}")
        print(f"  Violations: {len(self.compliance_violations)}")
        print(f"  Session: {self.session_id}")
        
        if self.conversation_history:
            avg_score = sum(entry["compliance_score"] for entry in self.conversation_history) / len(self.conversation_history)
            print(f"  Avg Compliance Score: {avg_score:.1%}")
    
    def _show_violations(self):
        """Show compliance violations"""
        if not self.compliance_violations:
            print("✅ No compliance violations detected")
            return
        
        print(f"\n⚠️  Compliance Violations ({len(self.compliance_violations)}):")
        for i, violation in enumerate(self.compliance_violations, 1):
            print(f"  {i}. [{violation['severity']}] {violation['timestamp']}")
            print(f"     PII Items: {violation['pii_count']}")
            print(f"     Original: {violation['original_message'][:50]}...")
            print(f"     Redacted: {violation['redacted_message'][:50]}...")
    
    def _show_final_summary(self):
        """Show final demo summary"""
        print("\n" + "=" * 50)
        print("📊 DEMO SUMMARY")
        print("=" * 50)
        print(f"Session ID: {self.session_id}")
        print(f"Total Messages: {len(self.conversation_history)}")
        print(f"Compliance Violations: {len(self.compliance_violations)}")
        
        if self.conversation_history:
            avg_score = sum(entry["compliance_score"] for entry in self.conversation_history) / len(self.conversation_history)
            print(f"Average Compliance Score: {avg_score:.1%}")
        
        print("\n🔧 TECHNICAL FEATURES DEMONSTRATED:")
        print("✅ Real-time PII detection")
        print("✅ Automatic redaction")
        print("✅ Compliance scoring")
        print("✅ Session tracking")
        print("✅ Brev LLM integration")
        print("✅ Multi-agent architecture")
        
        print("\n🚀 READY FOR JUDGES!")
        print("=" * 50)

def main():
    """Main entry point"""
    print("🤖 Starting AI Chatbot Compliance Demo...")
    
    # Check if Rust binary exists
    rust_binary = "../target/debug/pii-compliance-agent"
    if not os.path.exists(rust_binary):
        print(f"❌ Rust binary not found: {rust_binary}")
        print("Please build the Rust project first: cargo build")
        return
    
    # Check Brev configuration
    brev_api_key = os.getenv("BREV_API_KEY", "")
    if not brev_api_key:
        print("⚠️  BREV_API_KEY not set - using mock responses")
        print("Set BREV_API_KEY environment variable for real LLM responses")
    
    # Run demo
    demo = ChatbotComplianceDemo()
    demo.run_demo()

if __name__ == "__main__":
    main() 

File: src/utils/chatbot_demo.rs
use crate::types::{ChatMessage, ChatSession, RiskLevel};
use chrono::Utc;
use uuid::Uuid;

pub struct ChatbotDemoData;

impl ChatbotDemoData {
    pub fn new() -> Self {
        Self
    }
    
    pub fn get_chat_scenarios(&self) -> Vec<ChatSession> {
        vec![
            // Scenario 1: Customer Support with PII
            self.create_customer_support_scenario(),
            
            // Scenario 2: Healthcare Chat with Sensitive Data
            self.create_healthcare_scenario(),
            
            // Scenario 3: Banking Chat with Financial Data
            self.create_banking_scenario(),
            
            // Scenario 4: Clean Conversation
            self.create_clean_scenario(),
        ]
    }
    
    fn create_customer_support_scenario(&self) -> ChatSession {
        let session_id = Uuid::new_v4().to_string();
        let user_id = "user_123".to_string();
        
        let messages = vec![
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "Hi, I need help with my account".to_string(),
                is_user_message: true,
            },
            ChatMessage {
                user_id: "bot_001".to_string(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "Hello! I'd be happy to help. Can you provide your email address?".to_string(),
                is_user_message: false,
            },
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "Sure, my email is john.doe@gmail.com and my phone is 555-123-4567".to_string(),
                is_user_message: true,
            },
        ];
        
        ChatSession {
            session_id,
            user_id,
            messages,
            compliance_violations: vec![],
            risk_level: RiskLevel::Medium,
        }
    }
    
    fn create_healthcare_scenario(&self) -> ChatSession {
        let session_id = Uuid::new_v4().to_string();
        let user_id = "user_456".to_string();
        
        let messages = vec![
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "I need to schedule an appointment".to_string(),
                is_user_message: true,
            },
            ChatMessage {
                user_id: "bot_002".to_string(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "I can help you schedule an appointment. What's your name and date of birth?".to_string(),
                is_user_message: false,
            },
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "My name is Jane Smith, DOB 05/15/1985, SSN 123-45-6789".to_string(),
                is_user_message: true,
            },
        ];
        
        ChatSession {
            session_id,
            user_id,
            messages,
            compliance_violations: vec![],
            risk_level: RiskLevel::Critical,
        }
    }
    
    fn create_banking_scenario(&self) -> ChatSession {
        let session_id = Uuid::new_v4().to_string();
        let user_id = "user_789".to_string();
        
        let messages = vec![
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "I need to update my payment method".to_string(),
                is_user_message: true,
            },
            ChatMessage {
                user_id: "bot_003".to_string(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "I can help you update your payment method. What's your card number?".to_string(),
                is_user_message: false,
            },
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "My card number is 4111-1111-1111-1111, exp 12/25".to_string(),
                is_user_message: true,
            },
        ];
        
        ChatSession {
            session_id,
            user_id,
            messages,
            compliance_violations: vec![],
            risk_level: RiskLevel::High,
        }
    }
    
    fn create_clean_scenario(&self) -> ChatSession {
        let session_id = Uuid::new_v4().to_string();
        let user_id = "user_999".to_string();
        
        let messages = vec![
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "What's the weather like today?".to_string(),
                is_user_message: true,
            },
            ChatMessage {
                user_id: "bot_004".to_string(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "I can check the weather for you. What city are you in?".to_string(),
                is_user_message: false,
            },
            ChatMessage {
                user_id: user_id.clone(),
                session_id: session_id.clone(),
                message_id: Uuid::new_v4().to_string(),
                timestamp: Utc::now(),
                content: "I'm in San Francisco".to_string(),
                is_user_message: true,
            },
        ];
        
        ChatSession {
            session_id,
            user_id,
            messages,
            compliance_violations: vec![],
            risk_level: RiskLevel::Safe,
        }
    }
} 